### PR TITLE
Add the ability to create a new page in the Site Editor

### PR DESCRIPTION
**Pull request by SaxonF**
## What?
As part of https://github.com/WordPress/gutenberg/issues/44461 we are re-introducing a way to edit content within the site editor. Adding new content, such as pages, is an important part of that experience. This PR gives us a very basic starting experience to work from. 

## Why?
Resolves first iteration of https://github.com/WordPress/gutenberg/issues/48456

## How?
The AddNewPageModal sits above site view / editor and adds open state to store so that it can be called upon in different places (e.g. navigation block). In future this might be more appropriate as a package.

TODO: We'll want the command center "new page" command to make use of this modal as well.

## Testing Instructions
1. Open site editor
2. Click on pages
3. Click the `+` button
4. Notice new page modal opens
5. Give page a title and hit create
6. You should be redirected to editor with new page loaded

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/1072756/596edf2e-9756-456b-8b26-b2dabea28828

### CODE REVIEW

**Comment by jameskoster** [comment]
This looks really good to me 🚀

Should there be an indication somewhere that the new page will be a draft? Modal title could be "Draft a new page", or there could be a line of copy therein. The Snackbar is another option. 

I'm a bit on the fence about merging, given the current tension between content/template in edit view. It's not so bad for existing pages because the content is easier to locate. But for new pages with empty post content it's very easy to start adding content to the template. Should we tackle https://github.com/WordPress/gutenberg/issues/49086 first? Alternatively, could new pages created this way inject some example copy into Post Content for the user to replace?

**Comment by mtias** [comment]
We should have a component that doesn't redirect by default. In some context, it'd be fine to open the page as you are creating it, but in others it'd be a bit jarring — for example, from Link UI in menus or button blocks. Though we would want to be able to use the same modal.

I really think we should try to have the parent selector here in an unobtrusive way, default to "top level".

**Comment by SaxonF** [comment]
> We should have a component that doesn't redirect by default. In some context, it'd be fine to open the page as you are creating it, but in others it'd be a bit jarring

I've adjusted action to include option to redirectAfterSave though implementation using store is a little weird.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: Why is this using a store value, instead of just rendering the the `AddNewPageModal` here and using local state only render `AddNewPageModal` when it's open?

Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: I can change to do that if you think it's a better approach. Was thinking about other triggers like Link UI or command center.

Comment (youknowriad) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: LinkUI is in the block editor package and command center is in its own package too. I'm not sure how exactly they can share the same UI. It might be possible but there's a lot of challenges there. So I think to start, it's better to avoid any new API.

Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: Sounds good. I'll update tomorrow.

Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: Updated to use local state and onSave.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: Why do we need to both set page and push history, and I think the setPage can probably be removed.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js:101]: Can we keep this callback within the modal component maybe?

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: in other modal components, this prop is named `onClose` should we do the same?

**Review by youknowriad — COMMENTED** [review]
This is getting close, nice work.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js:101]: Went back and forth on this and ended up in this approach as it felt a little more flexible but happy to change if you'd prefer to move it in modal.

Comment (youknowriad) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js:101]: I'm not strongly opinionated either. I can see arguments for both.

File: packages/edit-site/src/components/add-new-page/index.js
Comment (scruffian) [inline packages/edit-site/src/components/add-new-page/index.js:86]: I think it would be better to have no default title and use a placeholder here.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (scruffian) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js:204]: Should we use `history` for this too?

Comment (scruffian) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js:204]: In fact why do we need an onClick at all?

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (scruffian) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: If the page title is an empty string this will return an empty string rather than `(no title`).

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js:204]: No idea. I haven't touched the "manage all pages" behaviour as part of this PR. Could we create a follow up issue? I know the pages list is being worked on as part of https://github.com/WordPress/gutenberg/issues/50418

File: packages/edit-site/src/components/add-new-page/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/add-new-page/index.js:86]: Removed default and using placeholder now. Thanks @scruffian !

File: packages/edit-site/src/components/add-new-page/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-page/index.js]: the title needs to be translated (wrapped in `_( )` )

File: packages/edit-site/src/components/add-new-page/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-page/index.js]: These two props (label and help) need to be translated too.

**Comment by youknowriad** [comment]
Aside the two translation comments, this is looking good to me. 
Ideally, we'd also add e2e tests

**Comment by SaxonF** [comment]
> Ideally, we'd also add e2e tests

Resolved translation issues and added an e2e test

**Comment by youknowriad** [comment]
While testing this PR, I noticed a weird bug:

 - After creating a page
 - There's a tooltip that is visible on the left and that's because there's a "hidden" button that is focused. The issue seems to be that the "sidebar" is not actually removed from the DOM but just "hidden" and I don't think that's a good thing. It was probably done that way in some animation refactoring or something. Not sure if you know anything about that.

Anyway, that bug is unrelated and I think this is ready to merge after rebase...

**Comment by SaxonF** [comment]
@youknowriad  I can look more into it on Monday but I did a little bit of digging. I could be way off base here but the below shows the sidebar when in edit mode with `opacity: 0;` , in code it should be removing from DOM but it looks like there might be an issue with `AnimatePresence` that's triggering the exit animation but not removing it. I think we updated Framer Motion at some point cc @ciampo 

image

image

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: I think this should be `(no title)` because that's what we used everywhere in WordPress.

**Review by mirka — COMMENTED** [review]
> it should be removing from DOM but it looks like there might be an issue with `AnimatePresence` that's triggering the exit animation but not removing it. I think we updated Framer Motion at some point cc @ciampo

The `history.push()` here seems to be throwing it off. Calling `setCanvasMode()` explicitly (like in the [other transitions](https://github.com/WordPress/gutenberg/blob/b3ea2b6a80ee92f26148899455669c53ae2f49de/packages/edit-site/src/components/block-editor/editor-canvas.js#L63)) triggers the sidebar unmount properly.

```diff
diff --git a/packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js b/packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
index 3e2c22fb01..44f6440a89 100644
--- a/packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
+++ b/packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
@@ -12,6 +12,7 @@ import { useEntityRecords } from '@wordpress/core-data';
 import { decodeEntities } from '@wordpress/html-entities';
 import { plus } from '@wordpress/icons';
 import { privateApis as routerPrivateApis } from '@wordpress/router';
+import { useDispatch } from '@wordpress/data';
 
 /**
  * Internal dependencies
@@ -23,6 +24,7 @@ import SidebarNavigationSubtitle from '../sidebar-navigation-subtitle';
 import SidebarButton from '../sidebar-button';
 import AddNewPageModal from '../add-new-page';
 import { unlock } from '../../private-apis';
+import { store as editSiteStore } from '../../store';
 
 const { useHistory } = unlock( routerPrivateApis );
 
@@ -42,10 +44,13 @@ export default function SidebarNavigationScreenPages() {
 	);
 
 	const [ showAddPage, setShowAddPage ] = useState( false );
+	const { setCanvasMode } = unlock( useDispatch( editSiteStore ) );
 
 	const history = useHistory();
 
 	const handleNewPage = ( { type, id } ) => {
+		setCanvasMode( 'edit' );
+
 		// Navigate to the created template editor.
 		history.push( {
 			postId: id,
```

**Comment by youknowriad** [comment]
I think instead of calling `setCanvas` we should try to figure out why the history change is not doing it properly.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: @scruffian I think this is something we can address [in this issue](https://github.com/WordPress/gutenberg/pull/50630) as its more directly related to the pages list

**Comment by SaxonF** [comment]
I think there are a couple issues going on here. 

The first was that in `use-sync-canvas-mode-with-url.js` we were setting currentCavasMode before performing checks, which meant it was always equal to canvasInUrl (see below). I've moved the setting below the checks.

```
useEffect( () => {
		currentCanvasInUrl.current = canvasInUrl;
		if (
			canvasInUrl === undefined &&
			currentCanvasMode.current !== 'view'
		) {
			setCanvasMode( 'view' );
		} else if (
			canvasInUrl === 'edit' &&
			currentCanvasMode.current !== 'edit'
		) {
			setCanvasMode( 'edit' );
		}
	}, [ canvasInUrl, setCanvasMode ] );
```

The second is that I think there is a bug with AnimatePresence (most likely [here](https://github.com/framer/motion/pull/1569) and [here](https://github.com/framer/motion/issues/1624)).

You can see in the logging below that `showSidebar` is being set correctly, and the exit animation is firing, but the sidebar isn't being unmounted. I came across a similar issue in the frame resizing work when trying to perform an exit animation on the resize handler.  

image

@mirka I'm assuming this will need to solved at the component level ideally.

**Comment by carolinan** [comment]
Not a code review **and not blocking:**
Creating a new page this way feels fast, and I like it.
The texts clearly says that I am creating a draft, and that is important.

But I still find the saving and publishing flow confusing. Once I have added a page title in the small page creation modal and selected Create draft, I am sent to the combined template and page editor, and it doesn't look like the page is in draft anymore, because the Save button is disabled. It is also confusing that it is not a "publish" button. 
I am also not asked if I want to use a pattern to get started with my content.

File: packages/edit-site/src/components/add-new-page/index.js
Comment (carolinan) [inline packages/edit-site/src/components/add-new-page/index.js]: Minor, not blocking: WordPress does not mention this when creating a page using the "old" method. I am not sure it is needed.

Comment (SaxonF) [reply packages/edit-site/src/components/add-new-page/index.js]: Thanks @carolinan ! I've removed the help text.

**Comment by SaxonF** [comment]
> But I still find the saving and publishing flow confusing. 

This is good feedback. We have a [separate issue](https://github.com/WordPress/gutenberg/issues/49873) for handling publish flow and some [follow up work](https://github.com/WordPress/gutenberg/issues/48456) around adding pages that includes pattern insertion.

**Comment by mirka** [comment]
I haven't gotten to the root of it, but I'll add some debugging notes since I'm going to be away for two weeks.

> The first was that in `use-sync-canvas-mode-with-url.js` we were setting currentCavasMode before performing checks, which meant it was always equal to canvasInUrl (see below). I've moved the setting below the checks.

I'm afraid this doesn't change anything. It's easy to misread because of the variable naming, but the `currentCanvasInUrl` ref is only read in the _other_ `useEffect`, so it doesn't affect the checks in this `useEffect`.

> The second is that I think there is a bug with AnimatePresence (most likely [here](https://github.com/framer/motion/pull/1569) and [here](https://github.com/framer/motion/issues/1624)).

These AnimatePresence bugs are interesting. Though, it might be entirely unrelated, because when you log everything the `showSidebar` variable fluctuates _even more_ in our success case. This is because `canvasMode` is actually set to `"view"` once before being set back to the correct `"edit"`, due to location params being stale until the next render. (Not sure if this is intentional, though theoretically it shouldn't be a problem as long as they're eventually correct.)

## Logs

<details><summary>✅ Success case (whenever <code>setCanvasMode()</code> is called before the <code>history.push()</code>, i.e. the useEffects are first triggered by a <code>setCanvasMode()</code>)</summary>
<p>
Console logs of what happens in the success case
</p>
</details> 

<details><summary>❌ Failure case (whenever the <code>history.push()</code> is called before <code>setCanvasMode()</code>, i.e. the useEffects are first triggered by a <code>history.push()</code>)</summary>
<p>
Console logs of what happens in the failure case
</p>
</details> 

---

Next steps for me would be to get a more minimal repro in a sandbox, but depending on what your time/resource constraints are for this PR, the quick workaround is to call `setCanvasMode()` first.

**Comment by youknowriad** [comment]
> This is because canvasMode is actually set to "view" once before being set back to the correct "edit", due to location params being stale until the next render. 

I guess that's because we have two separate hooks that read the "url" and update the state (one for canvas, one for the sidebar path). It's not entirely on purpose. Ideally we'd have only one hook to perform the `path <=> state` synchronization but I'm not sure how possible is that with our existing components.

**Comment by youknowriad** [comment]
I think the workaround (setting canvas) might have another unwanted side effect: "creating two browser navigation steps" you'd have to click "previous" twice to go back to the previous page (to be tested though). That said, if the workaround works without issues, let's add an inline comment to explain why we had to do it.

File: packages/edit-site/src/components/add-new-page/index.js
Comment (jeryj) [inline packages/edit-site/src/components/add-new-page/index.js:79]: Could we focus this input automatically when the modal is opened?

File: packages/edit-site/src/components/add-new-page/index.js
Comment (jeryj) [inline packages/edit-site/src/components/add-new-page/index.js:50]: After the page is created, where should focus get placed? Right now it's being dropped at the hidden Back button in the closed sidebar. I agree with @carolinan that the current mode of being moved to the site editor template for this page is an odd experience. I'd be fine with it if the sidebar stayed open, in which case, handling the focus here isn't an issue since it's being placed in a sensible spot.

**Comment by SaxonF** [comment]
Thanks y'all, I'll test setting canvas today. 

@mirka thanks for pointing out the variable names, my brain somehow couldn't distinguish between them 😆

File: packages/edit-site/src/components/add-new-page/index.js
Comment (SaxonF) [reply packages/edit-site/src/components/add-new-page/index.js:50]: >  the current mode of being moved to the site editor template for this page is an odd experience

This will make more sense once the [content editing PR](https://github.com/WordPress/gutenberg/pull/50857) is merged. There is also some [follow up work](https://github.com/WordPress/gutenberg/issues/48456) around improving the add page experience, including selecting starter patterns.

**Comment by SaxonF** [comment]
Unfortunately I don't think adding setCanvasMode before or after the history push is going to work without also looking at `use-sync-canvas-mode-with-url.js `. As @youknowriad touched on, the history push in sync-canvas executes last (for some reason) with stale params so you're always taken to `path=%2Fpage&canvas=edit` instead of your new page.

I'd be surprised if this wasn't an issue with `AnimatePresence`. The sidebar is removed from DOM properly if you remove AnimatePresence, and you can see the exit animation is firing. I also tried adding `key` to each child within AnimatePresence which they should have without success.

Since this is an issue that's already in production affecting template and template part creation, could we spin up a separate issue to tackle the problem more holistically?

File: packages/edit-site/src/components/add-new-page/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/add-new-page/index.js:83]: > Give page a title and hit create

Not a blocker, just asking if we're not enforcing a page title for creation, maybe add `'Page title (optional)'`?

File: packages/edit-site/src/components/add-new-page/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/add-new-page/index.js]: Should this match the "No title" placeholder in the [post title block](https://github.com/WordPress/gutenberg/blob/ef1c31c2dd514372c783315b70eaa7f014aa2775/packages/block-library/src/post-title/edit.js#L70-L70)?

Screenshot 2023-06-06 at 11 38 14 am

File: packages/edit-site/src/components/add-new-page/index.js
Comment (ramonjd) [reply packages/edit-site/src/components/add-new-page/index.js:79]:
```suggestion
					<TextControl
						/* eslint-disable jsx-a11y/no-autofocus */
						autoFocus
						/* eslint-enable jsx-a11y/no-autofocus */
```

Agree it would be nice to have the input focussed. It's done elsewhere across the app, but just checking with @tellthemachines if there are any major accessibility concerns.

File: packages/edit-site/src/components/add-new-page/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/add-new-page/index.js:76]: Not a blocker, but I noticed that when I throttled my connection I could close the modal and click around, only to be redirected to the new page, which seemed confusing.

A bit of an edge case though, unless you live in an area with bad internet speeds like me 😄 

https://github.com/WordPress/gutenberg/assets/6458278/6106bb8a-6508-4b73-9dfc-d4a3dcf35b2f

Maybe in another iteration we could look at blocking the modal close until the save callback is fired?

```js
	const closeModal = () => {
		if ( isCreatingPage ) {
			return;
		}
		onClose();
	};
```

**Review by ramonjd — APPROVED** [review]
This is working well for me. Good first iteration!

- A new page is created as expected
- Checked in Safari/Chrome/Firefox
- New E2E tests pass (I think the failing E2E on this branch is unrelated?)

I tend to agree with @carolinan about where I should be publishing the page, but I think we can iron out those details later?

I left some general comments and questions. Maybe matching the "No title" in the placeholder could be a quick win before merge?

File: packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js
Comment (getdave) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: > Was thinking about other triggers like Link UI or command center.

This is correct. And also `Navigation` section in Browse Mode.

> LinkUI is in the block editor package and command center is in its own package too. I'm not sure how exactly they can share the same UI.

The link control doesn't need to care about _how_ the create page happens. All it needs is a function it should trigger when someone clicks the appropriate piece of UI. That would be possible to pass to the `LinkControl` component via block editor settings in the same way we have had to do for many aspects of this component,

The other use case of the `Navigation` section of Browse Mode is even more clear cut as it sits adjacent to the `Pages` panel.

@youknowriad would you be open to be reverting this back to a global implementation in order that it can be triggered from elsewhere in the interface. I think "Create new page" is a pretty standard interaction for the editor as a whole.

Comment (getdave) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: > The link control doesn't need to care about how the create page happens. 

Actually one thing I realised is that the Link Control would need to have a way to get the data about the created page. That could be slightly tricky as it doesn't have direct access to Core Data.

@youknowriad The inability for key editor components to access Core Data has been a long standing issue. Is it not possible to provide access to these via some kind of proxy that we can pass into block editor settings? That way the block editor package can remain de-coupled but still have a means to access the kind of data it requires.

Comment (youknowriad) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js]: > The inability for key editor components to access Core Data has been a long standing issue. Is it not possible to provide access to these via some kind of proxy that we can pass into block editor settings?

It's only an issue if you think from a WordPress editor perspective. IMO, it's actually a good thing that force us to think about the right abstractions in the right place. Without these kind of constraints and abstractions, Gutenberg would not be usable in Tumblr, Drupal...

**Pull request merged**
